### DISCUSSION TITLE
Infinite recursion when using users table to specify users role for RLS

### DISCUSSION THREAD

**Discussion by solvemon-vh** [supabase/supabase#1138] [Questions] [answered]
I am trying to implement a relatively simple user system:

* Normal users can only view themselves
* Admin users can view, update or delete other users

As per suggested I have created a copy of the users table in public. Not the `is_admin` field:
```SQL
CREATE TABLE users
(
    id        uuid         not null constraint users_pkey primary key,
    name      varchar(255),
    email     varchar(255) not null,
    is_admin  boolean default false    
);   
```

any automatically copy from auth to public:
```SQL        
        CREATE FUNCTION public.handle_new_user()
        RETURNS TRIGGER AS $$
        BEGIN
        INSERT INTO public.users (id, name, email)
        VALUES (new.id, new.email, new.email);
        RETURN new;
        end;
        $$ LANGUAGE plpgsql security definer;

        CREATE TRIGGER on_auth_user_created AFTER INSERT ON auth.users
               FOR EACH ROW EXECUTE PROCEDURE public.handle_new_user();        
```

Now I try setting up the policies for normal users:
```SQL
CREATE POLICY "User can only view self"
        ON users FOR SELECT using (
            auth.uid() = id
        )
``` 
So far so good.

Now here comes the problem:
```SQL
CREATE POLICY "Full access on all users for admins"
        ON users FOR ALL using (
            auth.uid() in (select id from users where is_admin = true)            
        )
```
This causes an infinite recursion, because the policy lookup is also subject to the policies. 

One thing I did try was to create a view called `admin_users` and then use this to check if `auth.uid() in (select id from admin_users)`, however the view is then exposed on the API without RLS. If I change the owner of the view (to authenticated for example), the policy loses access to the view.

I am sort of stuck here, any suggestions on where to go from here?

**Comment by solvemon-vh** [comment, marked as answer]
Oh man, typical. Spend a whole day struggling with a problem, and find the solution 10 minutes after creating the thread :-/

My solution was to create a function defined as SECURITY DEFINER. This means the function is executed by the privilege of the creator, not the invoker, and thus will not be restricted. Furthermore, the function is not exposed to the API like the view, so I do not have to worry about that part either.

```SQL
CREATE OR REPLACE FUNCTION is_admin(user_id uuid)
            returns boolean AS
        $$
        select is_admin from users
        where id = user_id
        $$ stable language sql security definer;
```

```SQL
CREATE POLICY "Full access on all users for admins"
        ON users FOR ALL using (is_admin(auth.uid()))        
```

**Comment by ggimenes** [comment]
I just an issue here, but didn't pass the user_id as parameter for security concerns  [#6437](https://github.com/orgs/supabase/discussions/6467#discussioncomment-2582033)

```
CREATE OR REPLACE FUNCTION is_admin()
returns boolean
set search_path = public
AS $$
BEGIN
    RETURN (
        (select auth.jwt()) ->> 'role' = 'admin'
        OR
        EXISTS (
            SELECT 1 FROM auth.users
            WHERE id = (select auth.uid())
            AND raw_user_meta_data->>'role' = 'admin'
        )
    );
END;        
$$ STABLE LANGUAGE plpgsql SECURITY DEFINER;

-- Create new policies with proper authentication and profile-based checks
CREATE POLICY "Users and Admins can view their own positions"
ON "public"."jobs"
FOR SELECT
TO authenticated
USING ((SELECT auth.uid()) = user_id OR is_admin());
```
